Here's a pre-spec

lgierth opened the pull request:
- Q: how did you draw the ascii art?
- A: by hand, with love

lgierth commented:
Rendered view: https://github.com/libp2p/go-libp2p-relay/blob/feat/prespec/README.md

lgierth commented:
The section on addressing is still a bit rough around the edges, but I think it all captures the general idea.

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
+[go-libp2p-nat]: https://github.com/libp2p/go-libp2p-nat
+
+
+## Dramatization
```
❤️

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
+This helps in situations where nodes are behind NAT or reverse proxies,
+or simply don't support the same transports (e.g. go-ipfs vs. browser-ipfs).
+libp2p already has modules for NAT ([go-libp2p-nat][go-libp2p-nat]),
+but these don't always do the job, just because NAT traversal is a clusterfuck.
```
> clusterfuck

technical term :D

we'll want to remove before merging for benefit of more proper readers

lgierth replied:
yes, did this in the spirit of banana-driven writing. if you can't come up with a good term, insert one which is so bad you *have* to replace it :)

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
+One node asks a relay node to connect to another node on behalf of the first node.
+The relay node shortcircuits its streams to the two nodes,
+and they are then connected through the relay.
+This connection quaks and behaves just like a direct connection would,
```
rm "quaks and"

jbenet replied:
:duck: :duck:

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
++-------+                                        +---------+                               +-------+
+      ^                                            +-----+                                    ^
+      |                                            |     |                                    |
+      |        /libp2p-circuit-relay/QmTwo         |     |                                    |
```
can we come up with a shorter protocol name? like `p2p-circuit` or `p2pcr` or something?

lgierth replied:
I'm open to anything here -- in a similar fashion we already have `/libp2p-webrtc-star` and `/libp2p-webrtc-direct`

diasdavid replied:
yeah, we can short all the libp2p specific transports to only use `p2p`, at the same time we drop `/ipfs/` and use `/p2p/` for addrs

lgierth replied:
/p2p-circuit then?

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
+- QmOne is configured to use QmRelay for relaying.
+- QmOne automatically added `/libp2p-circuit-relay/ipfs/QmTwo` to its set of QmTwo addresses.
+- QmOne tries to connect via relaying, because it shares this transport with QmTwo.
+- A lively and prolonged dialogue ensues.
```
expand with:

Scene 3:
- QmOne connects to QmRelay, establishing a secure connection.
- QmOne asks QmRelay to connect to `/ipfs/QmTwo`
- QmRelay can do so, and dials QmTwo, establishing a secure connection.
- QmRelay then plugs the connections together, letting QmOne speak to QmTwo.
- QmOne and QmTwo want to speak privately, so they, too, establish a secure connection over the relayed transport.
- A lively and secure dialogue ensues.

diasdavid replied:
👍  Yeah, this is why we need Relay as a Mounted protocol

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
+  - need to update the multiaddr size from V to 0
+Q: should the relay transport be configured a a simple listener?
+  - e.g. simply on /libp2p-circuit-relay
+  - or should it just be enabled by default?
```
dont understand the Q

diasdavid replied:
I believe the question/answer is that we need to migrate Swarm addrs to be:
- addrs to listen
- addrs to announce
- addrs to not announce

Relay would be an addr to announce

jbenet commented on README.md:
````diff
@@ -1,2 +1,175 @@
+
+```
+<multistream-header>
+<src>
````
maybe say `<src-multiaddr>`

lgierth replied:
Yes these should be multiaddrs

jbenet commented on README.md:
````diff
@@ -1,2 +1,175 @@
+
+```
+/ipfs/relay/circuit/0.1.0
+QmOne
````
`/ipfs/QmOne` ?

and it should be `/p2p` once we make that switch

lgierth replied:
You're right, changing it to `/p2p`

jbenet commented on README.md:
````diff
@@ -1,2 +1,175 @@
+```
+<multistream-header>
+<src>
+<dst>
````
`<dst-multiaddr>` ?

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
+- `multistream-header`: This field is not strictly part of the relay stream.
+  - Multistreams are prefixed with a multicodec field identifying the protocol used for the stream.
+- `src`: The PeerID of the node opening the connection.
+- `dst`: The PeerID of the node that `src` wants to connect to.
```
hmmm i may be fine if this is a peer-only thing. then it works differently than the circuit relay thing i had made, which dials to multiaddrs. (i.e. i tell A what to dial to connect to the next step)

jbenet commented on README.md:
````diff
@@ -1,2 +1,175 @@
+
+  AllowRelaying(bool whetherToRelayOthersConnections)
+}
+```
````
I cant comment on the interfaces because i havent used them :)

jbenet commented on README.md:
```diff
@@ -1,2 +1,175 @@
+- *Not capable of *connecting* via relaying.*
+
+Since the existing protocol is incomplete, insecure, and certainly not used,
+we can safely remove it.
```
Hm, we do need to be able to dial multiaddrs that the relaying node does not previously know. I think this is two things.

- (1) a circuit relay that dials other multiaddrs.
- (2) a circuit relay that understands `/{ipfs, p2p}/<peerid>` multiaddrs, and can dial to those with its own routing.

it would be useful to setup a bunch of boxes that do (1) kind of relay without dht or anything else. just dumb pipes.

lgierth replied:
Yes agreed, in the addressing section i wrote:

> Relay addresses can be encapsulated "infinitely",
> so that one can construct onion-like relay chains:
> `/libp2p-circuit-relay/ipfs/QmFoo/libp2p-circuit-relay/ipfs/QmBar`
> 
> Relay addresses can encapsulated in an /ipfs address to signal a specific relay node to use:
> `/ipfs/QmRelay/libp2p-circuit-relay/ipfs/QmFoo`
> or `/dns6/relay.libp2p.io/ipfs/QmRelay/libp2p-circuit-relay/ipfs/QmFoo`

the stream protocol could understand any of these.

diasdavid commented on README.md:
```diff
@@ -1,2 +1,175 @@
-# go-libp2p-relay
-Circuit Switching for libp2p
+# go-libp2p-circuit-relay
+> Circuit Switching in libp2p
```
The spec should be a libp2p spec inside the specs folder, it applies to any language (or at least, it should)

lgierth replied:
agreed, gonna move it there eventually

diasdavid replied:
Thank you for moving it :)

diasdavid commented on README.md:
````diff
@@ -1,2 +1,175 @@
+      +--------------------------------------------+     +------------------------------------+
+```
+
+[go-libp2p-nat]: https://github.com/libp2p/go-libp2p-nat
````
Why is this link here?

lgierth replied:
It's a footnote link, the actual link is within the paragraphs above. I'll inline it.

diasdavid commented on README.md:
```diff
@@ -1,2 +1,175 @@
+- QmOne wants to connect to QmTwo,
+  and through peer routing has acquired a set of addresses of QmTwo.
+- QmTwo doesn't support any of the transports used by QmOne.
+- Awkward silence.
```
ahaah 😁

diasdavid commented on README.md:
```diff
@@ -1,2 +1,175 @@
+or `/dns6/relay.libp2p.io/ipfs/QmRelay/libp2p-circuit-relay/ipfs/QmFoo`
+
+
+### /ipfs/relay/circuit stream protocol
```
can we do `/p2p-circuit-relay/` instead? `/ipfs/relay` relay would make it specific to IPFS and could also mean other types of relay

lgierth replied:
This is the protocol multicodec, not the multiaddr, as in: https://github.com/libp2p/specs/blob/master/7-properties.md#757-protocol-multicodecs

lgierth commented:
I think I've addressed all comments, plus the addressing and wire sections got a big makeover and are a lot better now. (I'm not sure yet how to fit in the "Scene 3" as suggested.)

dignifiedquire commented on README.md:
```diff
@@ -1,2 +1,200 @@
+
+## Addressing
+
+`/p2p-circuit` multiaddrs don't carry any meaning of their own.
```
Once we move `ipfs` to `p2p` this starts looking pretty odd having `p2p` twice in there.

lgierth replied:
I'm not sure how to avoid that:

1. multiaddrs are meant to be as verbose as neccessary.
2. some kind of prefix for libp2p-specific protocols (other example: `/libp2p-webrtc-star`) is likely helpful to open up development to other projects and communities.

I also don't mind it much really :)

dignifiedquire replied:
given that I don't have any good ideas I'm withdrawing my concerns :)

dignifiedquire replied:
given that I don't have any good ideas I'm withdrawing my concerns :)

diasdavid commented:
Humble request - Can this be moved to libp2p/specs?

lgierth commented:
Yes if it looks good to you? I just didn't wanna interrupt the PR in the middle of it

lgierth commented:
Got positive feedback from @whyrusleeping too -- closing this PR in favor of moving the document into libp2p/specs.
